File: python/src/algorithms/algorithm.py
from abc import ABC, abstractmethod
from typing import Any, Callable, Dict


class Algorithm(ABC):
    def __init__(self, supported_algorithms: Dict[str, Callable]):
        self.supported_algorithms = supported_algorithms

    @abstractmethod
    def generate_signature(
        self, encoded_header: str, encoded_payload: str, algorithm: str, key: Any
    ) -> str:
        pass

    @abstractmethod
    def verify_signature(
        self,
        encoded_header: str,
        encoded_payload: str,
        algorithm: str,
        signature: str,
        key: Any,
    ) -> bool:
        pass

    def is_algorithm_supported(self, algorithm: str) -> bool:
        return True if algorithm in self.supported_algorithms.keys() else False

File: javascript/src/algorithms/algorithm.js
export const Algorithm = function (supportedAlgorithms, constructor = null) {
  this.constructor = constructor === null ? this.constructor : constructor;
  if (this.constructor === Algorithm) {
    throw new Error("Abstract function");
  }

  this.supportedAlgorithms = supportedAlgorithms;

  this.generateSignature = (
    encodedHeader,
    encodedPayload,
    algorithm,
    secretKey
  ) => {
    throw new Error("Abstract property");
  };

  this.verifySignature = (
    encodedHeader,
    encodedPayload,
    algorithm,
    signature,
    secretKey
  ) => {
    throw new Error("Abstract property");
  };

  this.isAlgorithmSupported = (algorithm) => {
    for (const i of Object.keys(this.supportedAlgorithms)) {
      if (algorithm === i) {
        return true;
      }
    }
    return false;
  };
};
